tooling/minimize/src/enums.rs

```diff
@@ -0,0 +1,91 @@
+use crate::*;
+
+pub fn translate_enum<'tcx>(
+    ty: rs::Ty<'tcx>,
+    adt_def: rs::AdtDef<'tcx>,
+    sref: rs::GenericArgsRef<'tcx>,
+    tcx: rs::TyCtxt<'tcx>,
+) -> Type {
+    let a = rs::ParamEnv::empty().and(ty);
+    let layout = tcx.layout_of(a).unwrap().layout;
+    let size = translate_size(layout.size());
+    let align = translate_align(layout.align().abi);
+
+    // The discriminant type in Rust is always isize which always gets translated as i64.
+    let discriminant_ty = IntType { signed: Signedness::Signed, size: Size::from_bytes_const(8) };
+
+    let (variants, discriminator) = match layout.variants() {
+        rs::Variants::Single { index } => {
+            let fields = translate_fields(layout.fields(), adt_def.variant(*index), sref, tcx);
+            ([Variant { ty: Type::Tuple { fields, size, align }, tagger: Map::new() }].into_iter().collect(), Discriminator::Known(0.into()))
+        },
+        rs::Variants::Multiple {
+            tag,
+            tag_encoding,
+            tag_field,
+            variants,
+        } => {
+            
+            // compute the offset of the tag for the tagger and discriminator construction
+            let tag_offset: Offset = translate_size(layout.fields().offset(*tag_field));
+            // compute the tag type (which will always be unsigned as we get the discriminant as u128).
+            let tag_ty = match tag.primitive() {
+                rs::Primitive::Int(ity, _signed) => IntType { signed: Signedness::Unsigned, size: translate_size(ity.size()) },
+                _ => panic!("enum tag has invalid primitive type"),
+            };
+
+            // translate the variants
+            let translated_variants = adt_def.variants().iter_enumerated().map(|(variant_idx, variant_def)| {
+                let fields = translate_fields(&variants[variant_idx].fields, &variant_def, sref, tcx);
+                let discr = adt_def.discriminant_for_variant(tcx, variant_idx);
+                let tag_bits = discr.val & (u128::MAX >> (128 - tag_ty.size.bits().try_to_u8().unwrap()));
+                let tagger = match tag_encoding {
+                    rs::TagEncoding::Direct => [(tag_offset, (tag_ty, Int::from(tag_bits)))].into_iter().collect(),
+                    rs::TagEncoding::Niche { .. } => todo!("Implement Niche-encoded tags for enums (Timon)"),
+                };
+                Variant { ty: Type::Tuple { fields, size, align }, tagger }
+            }).collect();
+
+            // build the discriminator.
+            let discriminator = match tag_encoding {
+                rs::TagEncoding::Direct => {
+                    let children = adt_def.discriminants(tcx)
+                        .map(|(idx, discr)| (Int::from(discr.val & (u128::MAX >> (128 - tag_ty.size.bits().try_to_u8().unwrap()))), Discriminator::Known(idx.index().into())))
```

Comment:
Instead of redoing the math here, wouldn't it be easier to build up the tagger and the discriminator together in the variant loop above? For direct tags, they are very similar.